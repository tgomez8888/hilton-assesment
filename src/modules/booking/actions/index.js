import ActionType from "../constants/actionType";

export function setRoomsValue(rooms) {
  return {
    type: ActionType.SET_ROOMS_VALUE,
    payload: { rooms }
  };
}

export function roomValueChange(id, name, value) {
  return {
    type: ActionType.ROOM_VALUE_CHANGE,
    payload: { id, name, value }
  };
}

export function roomSelectionChange(id, name, value) {  
  return {
    type: ActionType.ROOM_SELECTION_CHANGE,
    payload: { id, name, value }
  };
}
