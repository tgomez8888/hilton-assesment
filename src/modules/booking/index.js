export {Booking} from "./components/Booking";
export {reducer} from "./reducers";
