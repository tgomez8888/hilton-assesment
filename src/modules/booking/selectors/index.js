import { keys, path, pipe, prop, sort } from "ramda";

const localizeState = path(["app", "booking"]);

export function getRoomsId(state) {
  return pipe(
    localizeState,
    prop("rooms"),
    keys,
    sort
  )(state);
}

export function getRoomById(state, id) {
  return pipe(
    localizeState,
    prop("rooms"),
    prop(id)
  )(state);
}
