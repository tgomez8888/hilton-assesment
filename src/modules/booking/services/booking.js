import {
  __,
  always,
  assoc,
  curry,
  evolve,
  gte,
  ifElse,
  identity,
  indexBy,
  lt,
  map,
  pipe,
  prop,
  values
} from "ramda";

export const defaultRooms = [
  { id: 1, priority: 1, selected: true, adults: "1", children: "0" },
  { id: 2, priority: 2, selected: false, adults: "1", children: "0" },
  { id: 3, priority: 3, selected: false, adults: "1", children: "0" },
  { id: 4, priority: 4, selected: false, adults: "1", children: "0" }
];

export const updateRoom =  curry((newRoom, rooms) => pipe(
  indexBy(prop("id")),
  assoc(newRoom.id, newRoom),
  values
)(rooms)); 

const isHigherPriority = curry((newRoom, room) => pipe(
  prop("priority"),
  lt(__, newRoom.priority)
)(room));

const isLowerOrSamePriority = curry((newRoom, room) => pipe(
  prop("priority"),
  gte(__, newRoom.priority)
)(room));

const setDefaultValues = evolve({
  selected: always(false),
  adults: always("1"),
  children: always("0")
});

/**
 * This method receives a room with changes and calculates if the other rooms need to changes.
 * If room is selected all rooms with higher priority (lower priority value) will 
 * be selected.
 * If room is not selected all rooms with lower or same priority (higher priority value) will be 
 * unselected and set default values.
 * @param {Object} newRoom This is the modified room object 
 * @param {[Object]} rooms The current list of rooms in state
 * @returns {[Object]} New list of rooms with modifications 
 */
export const processRoomSelectionChanges = (newRoom, rooms) => {
   
  const turnOnHigherPriority = map(
    ifElse(isHigherPriority(newRoom), assoc("selected", true), identity)
  ); 
  const turnOffLowerPriority = map(
    ifElse(
      isLowerOrSamePriority(newRoom),
      setDefaultValues,
      identity
    )
  );

  return  pipe(
    updateRoom(newRoom),
    ifElse(() => newRoom.selected, turnOnHigherPriority, turnOffLowerPriority)
  )(rooms);
}