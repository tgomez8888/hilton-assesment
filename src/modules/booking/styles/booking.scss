@import "./modules/variables";

body {
	font-size: 14px;
}

input, select, button {
	font-size: 14px;
	margin: 0;
}

.actions-panel {
	padding: 10px;
}

.card-container {
	background-color: $color-background;
	border: 4px solid $color-border;
	display: inline-block;
	margin: 10px;
	border-radius: 5%;
	.card-header {
		background-color: $color-background;
		font-weight: bold;
		padding: 5px;
	}
	.card-body {
		background-color: $color-background-content;
		padding: 5px;
	}
	.column {
		display: inline-block;
		padding-right: 5px;
	}	
}

.card-container.disabled {
	background-color: $color-background-disabled;
	border-color: $color-border-disabled;

	.card-header {
		background-color: $color-background-disabled;
		font-weight: normal;
	}

	.card-body {
		background-color: $color-background-disabled;
	}	
}

.guess-selector {
	label{
		font-size: 12px;		
	}	
	width: 50px;	
}