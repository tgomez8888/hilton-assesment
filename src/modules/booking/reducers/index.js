import { assoc, indexBy, pipe, prop } from "ramda";
import createReducer from "../../../utils/createReducer";
import ActionType from "../constants/actionType";
import { defaultRooms, processRoomSelectionChanges } from "../services/booking";

const initialState = {
  rooms: indexBy(prop("id"), defaultRooms)
};

const setRoomsValue = (state, { payload }) => {
  const { rooms } = payload;

  return assoc("rooms", indexBy(prop("id"), rooms), state);
};

const roomSelectionChange = (state, { payload }) => {
  const { id, name, value } = payload;
  const newRoom = pipe(
    prop(id),
    assoc(name, value)
  )(state.rooms);
  const newRooms = pipe(
    assoc(id, newRoom),
    processRoomSelectionChanges(newRoom)
  )(state.rooms);

  return assoc("rooms", newRooms, state);
};

const roomValueChange = (state, { payload }) => {
  const { id, name, value } = payload;
  const newRoom = pipe(
    prop(id),
    assoc(name, value)
  )(state.rooms);
  const newRooms = assoc(id, newRoom, state.rooms);

  return assoc("rooms", newRooms, state);
};

const handlers = {
  [ActionType.SET_ROOMS_VALUE]: setRoomsValue,
  [ActionType.ROOM_SELECTION_CHANGE]: roomSelectionChange,
  [ActionType.ROOM_VALUE_CHANGE]: roomValueChange
};

export const reducer = createReducer(initialState, handlers);

export default reducer;
