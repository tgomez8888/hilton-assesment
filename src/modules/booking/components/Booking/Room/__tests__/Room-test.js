import { mount } from "enzyme";
import EnhancedRoom, { Room } from "..";
import { testComponent } from "../../../../../../utils/test/testComponent";
import { defaultRooms } from "../../../../services/booking";

describe("Room", () => {
  function render({ room }) {
    return mount(testComponent(Room, {room}));
  }

  function renderEnhanced({ id }) {
    return mount(testComponent(EnhancedRoom, {id}, {booking: {rooms: defaultRooms}}));
  }  

  test("renders correctly", () => {
    const room = {
      id: 1,
      priority: 2,
      selected: false,
      adults: 1,
      children: 0
    };
    const wrapper = render({ room });

    expect(wrapper.html()).toMatchSnapshot();
  });  

  test("loads the correct room from state and renders correctly", () => {    
    const wrapper = renderEnhanced({ id: 2 });

    expect(wrapper.html()).toMatchSnapshot();
  });  
});
