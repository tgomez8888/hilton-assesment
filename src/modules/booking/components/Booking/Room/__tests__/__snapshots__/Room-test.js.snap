// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room loads the correct room from state and renders correctly 1`] = `"<div class=\\"card-container disabled\\"><div class=\\"card-header\\"><input type=\\"checkbox\\" name=\\"selected\\">Room 3</div><div class=\\"card-body\\"><div class=\\"column guess-selector\\"><label>Adults (18+)</label><select name=\\"adults\\" disabled=\\"\\"><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div><div class=\\"column guess-selector\\"><label>Children (0-17)</label><select name=\\"children\\" disabled=\\"\\"><option value=\\"0\\">0</option><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div></div></div>"`;

exports[`Room renders correctly 1`] = `"<div class=\\"card-container disabled\\"><div class=\\"card-header\\"><input type=\\"checkbox\\" name=\\"selected\\">Room 1</div><div class=\\"card-body\\"><div class=\\"column guess-selector\\"><label>Adults (18+)</label><select name=\\"adults\\" disabled=\\"\\"><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div><div class=\\"column guess-selector\\"><label>Children (0-17)</label><select name=\\"children\\" disabled=\\"\\"><option value=\\"0\\">0</option><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div></div></div>"`;
